File: NumpyNetworks/trial.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>



double sigmoid(double x) {return 1 / (1 + exp(-x))}
double dSigmoid(double x) {return x * (1 - x)}

double init_weights() {return ((double)rand()) / ((double )) RAND_MAX)} // with random numbers

void shuffle(int *array, size_t n)
{
    if (n > 1)
    {
        size_t i;
        for (i = 0; i < n - 1; i++)
        {
            size_t j = i + rand() / (RAND_MAX / (n - i) + 1);
            int t = array[j];
            array[j] = array[i];
            array[i] = t;
        }
    }
}

#define numInputs 2  // number of input-nodes
#define numHiddenNodes 2
#define numOutputs 1
#define numTrainingSets 4  // number of examples

int main(void) {
    const double lr = 0.1f;

    double hiddenLayer[numHiddenNodes];  // 1D-list length of number of nodes in hidden-layer
    double outputLayer[numOutputs];     // 1D-list length of number of nodes in output-layer

    double hiddenLayerBias[numHiddenNodes];  // same as hidden-layer size
    double outputLayerBias[numOutputs];      // same as output-layer size

    double hiddenWeights[numInputs][numHiddenNodes];  // weights[prev-input][cur-hidden], w[input-node-i][hidden-node-i],weights from input-layer to hidden-layer
    double outputWeights[numHiddenNodes][numOutputs];  // weights[prev-hidden][cur-output], w[hidden-node-i][output-node-i],weights from hidden-layer to output=layer

    // train_x[number-of-examples][number-of-input-nodes]
    double training_inputs[numTrainingSets][numInputs] = {{0.0f, 0.0f},
                                                          {1.0f, 0.0f},
                                                          {0.0f, 1.0f},
                                                          {1.0f, 1.0f},
    }; // x-data 2D-list for each epoch you have each number of inputs per epoch. When you index a epoch you get a list of inputs for that epoch

    // train_y[number-of-examples][number-of-input-nodes]
    double training_outputs[numTrainingSets][numOutputs] = {{0.0f},
                                                          {1.0f},
                                                          {1.0f},
                                                          {0.0f},
    };
    
    // iterate number of inputs: initialize hidden-layer-weights
    for (int i=0; i<numInputs; i++) {
        // iterate number of hidden nodes
        for (int j=0; j<numHiddenNodes; j++) {
            // weights[input-node-i][hidden-node-in-layer], weights[prev-node-i][cur-node-i]
            hiddenWeights[i][j] = init_weights();  // get random weight value
        }
    }

    // iterate number of hidden nodes: initalize output-layer-weights: 
    for (int i=0; i<numHiddenNodes; i++) {
        // iterate number of output-nodes
        for (int j=0; j<numOutputs; j++) {
            // weights[inputs-i][hidden-node-in-layer], weights[prev-node-i][cur-node-i]
            outputWeights[i][j] = init_weights();
        }
    }

    // iterate number of nodes in hidden-layer: initalize hidden-layer bias
    for (int i=0; i< numHiddenNodes; i++) {
        hiddenLayerBias[i] = init_weights();
    }
    // iterate number of nodes in output-layer: initalize output bias
    for (int i=0; i< numOutputs; i++) {
        outputLayerBias[i] = init_weights();
    }


    int trainingSetOrder[] = {0,1,2,3};

    int numberOfEpochs = 10000;

    // training iterate each epoch
    for (int epoch=0; epochs < numberOfEpochs; epoch++) {
        shuffle(trainingSetOrder, numTrainingSets);
        // iterate number of numTrainingSets or examples 4
        for (int x=0; x<numTrainingSets; x++) {

            int i = trainingSetOrder[x];  // index of input

            // forward pass

            // iterate nodes in hidden-layer: compute hidden layer activation 1st-layer
            for (int j=0; j<numHiddenNodes; j++) {
                double activation = hiddenLayerBias[j];  // init activation sum to bias of jth node in hidden layer
                // iterate number of input nodes
                for (int k=0; k<numInputs; k++) {
                    // train_inputs[i-input-example-indx][k-input-node-indx] * hiddenWeights[k-input-node-indx][i-input-example-indx]
                    activation += training_inputs[i][k] * hiddenWeights[k][j];  // sum products of connections, hiddenWeights[prev][cur]
                }
                hiddenLayer[j] = sigmoid(activation);  // hiddenlayer[node-indx] acctivation
            }

            // iterate nodes in output-layer: compute output-layer-2 activations
            for (int j=0; j<numOutputs; j++) {
                double activation = outputLayerBias[j];   // init activation sum to bias of jth node in output-layer
                // iterate number of hidden-nodes 
                for (int k=0; k<numHiddenNodes; k++) {
                    // hiddenLayer[node-k] * outputWeights[hidden-node-k][output-node-j]
                    activation += hiddenLayer[k] * outputWeights[k][j];  
                }
                outputLayer[j] = sigmoid(activation);  
            }

            printf("Input: %g   Output: %g  Predicted: %g \n", training_inputs[i][0], training_inputs[i][0], training_inputs[i][0])

            // BACKPROPAGATION
            double deltaOutput[numOutputs];

            for (int j=0; j < numOutputs; j++) {
                double error = (training_outputs[i][j] - outputLayer[j]);
                deltaOutput[j] =  error * dSigmoid(outputLayer[j])
            }

            // compute chagne in hidden weights
            double deltaHidden[numHiddenNodes];
            for (j=0; j < numHiddenNodes; j++) {
                double error = 0.0f;
                for (int k=0; k <numOutputs; k++) {
                    error += deltaOutput[k] * outputWeights[j][k]
                }
                deltaHidden[j] = error * dSigmoid(hiddenLayer[j])
            }

            // Apply change in output weights
            for (int j=0; j < numOutputs; j++) {
                outputLayerBias[j] += deltaOutput[j] * lr;
                for (int k=0; k < numHiddenNodes; k++) {
                    outputWeights[k][j] += hiddenLayer[k] * deltaOutput[j] * lr;
                }
            }
            // Apply change in output weights
            for (int j=0; j < numHiddenNodes; j++) {
                hiddenLayerBias[j] += deltaHidden[j] * lr;
                for (int k=0; k < numInputs; k++) {
                    hiddenWeights[k][j] += train_inputs[i][k] * deltaHidden[j] * lr;
                }
            }
        }
    }
}


/* 

XOR Function:
input  |   ouput
0 0    |   0
1 0    |   1
0 1    |   1
1 1    |   0
if the 2 inputs are equal outputs 0, if they differ output 1.

Input nodes: 2
Hidden Layers: 1 with 2 nodes
Output nodes: 1




*/ 

File: NumpyNetworks/trial.py
import math
import random

def sigmoid(z):
    return 1 / (1+math.exp(-z))
def d_sigmoid(z):
    return z * (1-z)
def init_weight():
    return random.uniform(-1, 1)


num_inputs = 2
num_hidden_nodes = 2
num_outputs = 1
num_training_sets = 4   # number of examples

def main():

    lr = 0.1

    hidden_layer = []   # length of num-hidden-nodes
    output_layer = []   # length of num-outputs in output-layer

    hidden_layer_bias = []  # length of num-hidden-nodes
    output_layer_bias = []  # length of num-output-nodes in output-layer

    hidden_weights = []  # 2D: [input-layer-node-i][hidden-layer-node-i], connection from [prev-node]-to-[cur-node], [num_inputs][num_hidden_nodes]
    output_weights = []  # 2D: [hidden-layer-node-i][output-layer-node-i], connection from [prev-node]-to-[cur-node], [num_hidden_nodes][num_output_nodes]

    training_inputs = [[0.0, 0.0],  # 2D: [input-example-i][input-node-value-i], index what example and then the node of that example in input-layer
                       [1.0, 0.0],
                       [0.0, 1.0],
                       [1.0, 1.0]]
    
    training_outputs = [[0.0],  # 2D: [output-example-i][output-node-value-i], index what example and then the node of that example in output-layer
                       [1.0],
                       [1.0],
                       [0.0]]
    
    for i in range(num_inputs):     # iterate nodes in input-layer, no bias for input-layer
        hidden_weights.append([])
        for j in range(num_hidden_nodes):       # iterate nodes in hidden-layer
            hidden_weights[i].append(0)
            hidden_weights[i][j] = init_weight()        # init weight of hidden-weights[input-node-i][hidden-node-j]

    for i in range(num_hidden_nodes):       # iterate nodes in hidden-layer
        hidden_layer_bias.append(0)
        hidden_layer_bias[i] = init_weight()        # init bias for cur-node-i in hidden-layer
        output_weights.append([])
        for j in range(num_outputs):        # iterate nodes in output-layer
            output_weights[i].append(0)
            output_weights[i][j] = init_weight()    # init weight of output-weights[hidden-node-i][output-node-j]
        hidden_layer.append(0)

    for i in range(num_outputs):        # iterate nodes in output-layer
        output_layer_bias.append(0)
        output_layer.append(0)
        output_layer_bias[i] = init_weight()        # init bias for cur-node-i in output-layer 



    training_set_order = [0, 1, 2, 3]

    num_epochs = 10000      # number of iterations

    for epoch in range(num_epochs):     # iterate number of training epochs

        for x in range(num_training_sets):      # iterate 
            i = training_set_order[x]
            # FORWARD PASS

            # compute hidden-layer activation
            for j in range(num_hidden_nodes):       # iterate nodes in hidden-layer
                activation = hidden_layer_bias[j];      # init activation sum for cur-node-j in hidden-layer to bias of jth node in hidden-layer
                for k in range(num_inputs):     # iterate nodes in input-layer
                    activation += training_inputs[i][k] * hidden_weights[k][j]      # add to activation sum inputs[example-i][input-node-k] * hidden-weights[input-node-k][hidden-node-j]
                hidden_layer[j] = sigmoid(activation)       # set hidden-layer jth node activation to non-linearity of activation sum

            # compute output-layer activation
            for j in range(num_outputs):        # iterate nodes in output-layer
                activation = output_layer_bias[j]       # init activation sum cur-node-j in output-layer to bias of jth node in output-layer
                for k in range(num_hidden_nodes):       # iterate nodes in hidden-layer
                    activation += hidden_layer[k] * output_weights[k][j]    # add to activation sum # hidden-layer[node-k] * output-weights[hidden-node-k][output-node-j]
                output_layer[j] = sigmoid(activation)   # set output-layer jth node activation to non-linearity of activation sum


            print(f'Input: {training_inputs[i][0], training_inputs[i][1]}   Output: {output_layer[0]}   Expected: {training_outputs[i][0]}')
            # BACKPROPAGATION
                
            # compute change in each output-weight
            deltaOutput = []    # stores gradients of error with respect to nodes in output-layer, For each output node, it stores the product of the error and the derivative of the sigmoid activation function for that node.
            for j in range(num_outputs):        # iterate nodes in output-layer
                deltaOutput.append(0)
                errorOutput = training_outputs[i][j] - output_layer[j]      # compute difference in label-output[example-i][output-node-j] - output-layer[node-j]-activation
                deltaOutput[j] = errorOutput * d_sigmoid(output_layer[j])   # gradient of node-j in output-layer = error * sigmoid-derivative of output-layer[node-j
            
            # compute change in each hidden-weight
            deltaHidden = []    # stores gradients of error with respect to nodes in hidden layer
            for j in range(num_hidden_nodes):       # iterate nodes in hidden-layer
                deltaHidden.append(0)
                errorHidden = 0.0       #
                for k in range(num_outputs):
                    errorHidden += deltaOutput[k] * output_weights[j][k] 
                deltaHidden[j] = errorHidden * d_sigmoid(hidden_layer[j])

            # apply change in output-weights
            for j in range(num_outputs):
                output_layer_bias[j] = deltaOutput[j] * lr
                for k in range(num_hidden_nodes):
                    output_weights[k][j] += hidden_layer[k] * deltaOutput[j] * lr

            for j in range(num_hidden_nodes):
                hidden_layer_bias[j] += deltaHidden[j] * lr
                for k in range(num_inputs):
                    hidden_weights[k][j] += training_inputs[i][k] * deltaHidden[j] * lr
main()

"""
XOR Function:
input  |   ouput
0 0    |   0
1 0    |   1
0 1    |   1
1 1    |   0
if the 2 inputs are equal outputs 0, if they differ output 1.

Input nodes: 2
Hidden Layers: 1 with 2 nodes
Output nodes: 1



"""